sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.{DataFrame, QueryTest}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+
+class RemoveRedundantProjectsSuite extends QueryTest with SharedSparkSession {
+
+  private def assertProjectExecCount(df: DataFrame, expected: Integer): Unit = {
+    withClue(df.queryExecution) {
+      val plan = df.queryExecution.executedPlan
+      val actual = plan.collectWithSubqueries { case p: ProjectExec => p }.size
+      assert(actual == expected)
+    }
+  }
+
+  private def assertProjectExec(query: String, enabled: Integer, disabled: Integer): Unit = {
+    val df = sql(query)
+    assertProjectExecCount(df, enabled)
+    val result1 = df.collect()
+    withSQLConf(SQLConf.REMOVE_REDUNDANT_PROJECTS_ENABLED.key -> "false") {
+      val df2 = sql(query)
+      assertProjectExecCount(df2, disabled)
+      QueryTest.sameRows(result1.toSeq, df2.collect().toSeq)
+    }
+  }
+
+  private def withTestView(f: => Unit): Unit = {
+    withTempPath { p =>
+      val path = p.getAbsolutePath
+      spark.range(100).selectExpr("id % 10 as key", "id * 2 as a",
+        "id * 3 as b", "cast(id as string) as c", "array(id, id + 1, id + 3) as d")
+        .write.partitionBy("key").parquet(path)
+      spark.read.parquet(path).createOrReplaceTempView("testView")
```

Comment:
Shall we put the view creation in `beforeAll`? Then we only need to do it once for the entire test suite.